File: Prototype/BigGp/src/Model/BaseNodes/FunctionCallOut/FunctionCallOut.java
package Model.BaseNodes.FunctionCallOut;

import Model.BaseNodes.AddRandomBaseNode;
import Model.Interfaces.Crossable;
import Model.Interfaces.SubtreeMutable;
import Model.Node;
import Model.SmallNodes.Expressions.Standart.Childs.VarExpression;
import Serialization.Token;

import java.util.ArrayList;

public class FunctionCallOut extends Node implements Crossable, SubtreeMutable {

    private VarExpression variable;

    @Override
    public String toString() {
        return "write(" + variable.getVariableName() + ");";
    }

    public ArrayList<Node> getChildrenAsNodes(){
        ArrayList<Node> nodes = new ArrayList<>();
        nodes.add(this);
        nodes.add(variable);

        return nodes;
    }

    @Override
    public void subtreeMutate() {
        Node parentNode = this.getParentNode();
        double nextDeepNodeChance = ((AddRandomBaseNode) parentNode).getNextDeepNodeChance();
        Node node = ((AddRandomBaseNode) parentNode).getRandomNode(parentNode, nextDeepNodeChance);

        ((AddRandomBaseNode) parentNode).reAttachSubtree(this, node);
    }

    public FunctionCallOut(Node parent) {
        super(parent, "FunctionCallOut", true);

        variable = new VarExpression(this, true);
    }

    public FunctionCallOut(Node parent, ArrayList<Token> tokens) {
        super(parent, "FunctionCallOut", true);

        tokens.remove(0);
        variable = new VarExpression(this, tokens);
    }
}

File: Interactive.Genetic.Programming.Environment/src/Domain/Model/Nodes.Big/FunctionCall/FunctionCallOut.cs
using System.Collections;
using Model.Abstract;
using Model.Interfaces;
using Model.Nodes.Small.Expressions.Standard;

namespace Model.Nodes.Big.FunctionCall;

public sealed class FunctionCallOut : Node, ICrossable, ISubtreeMutable
{
    private readonly VarExpression _varExpression;

    public override List<Node> ChildrenAsNodes() => [this, _varExpression];

    public void SubtreeMutate()
    {
        var parent = ParentNode as BigNode ?? throw new InvalidOperationException("Parent is not a big node");

        var node = parent.GetRandomNode();
        parent.ReAttachSubtree(this, node);
    }

    public FunctionCallOut(Node parentNode) : base(parentNode, "FunctionCallOut", true)
    {
        _varExpression = new VarExpression(this, true);
    }
    
    public FunctionCallOut(Node parentNode, IList<Token> tokens) : base(parentNode, "FunctionCallOut", true)
    {
        tokens.RemoveAt(0);
        _varExpression = new VarExpression(this, tokens);
    }

    public override string ToString() => $"write({_varExpression});";
}